### PR TITLE
ability to build on mac, run an adapter

### LINKED ISSUE #495: create mac implementation of tun interface
Create a mac implementation of the tun interface.  Has to work with tokio too.

I would look at the rust WG code for inspiration.

### PR DESCRIPTION
**Pull request by mkolehmainen**
Permits compiling on a mac, and you can bring up an adapter and successfully connect to a node on linux.

There is still some work to do around the TUN interface.  This simply brings up an interface and sets an address (and mtu).  But we still need to add the correct routing.  

I'll make a story for that and for the "set MTU" thing for the UdpSocket stuff in zpr-ext.

### CODE REVIEW

File: adapter/ph/Cargo.toml [hunk 1]
```diff
@@ -23,16 +23,24 @@ openssl-sys = { version = "0.9.102" }
...
 thiserror = "1.0.65"
+zpr-ext = { path = "../../zpr-ext", features = ["bytes", "tokio", "zerocopy"] }
+
+[target.'cfg(target_os = "linux")'.dependencies]
+tokio-tun = { version = "0.11.5" }
+
+[target.'cfg(target_os = "macos")'.dependencies]
+# Locked verision of rust-tun as the maintainers may be suspect.
+# This is used for its ability to manage a TUN device on mac.
+tun = { version = "=0.7.1", features = ["async"] }
```
Comment (cpacejo) [inline adapter/ph/Cargo.toml:44]: 0.6.1 is the last "unadulterated" version.

Aside -- because we'll be open-source (as will our commit history), I'd reword the comment to be in neutral terms.  E.g. "Locked version, because rust-tun 0.7 was substantially reworked by a new maintainer, removing several features we depend on."

File: zpr-ext/src/tokio.rs [hunk 2]
```diff
@@ -2,12 +2,17 @@ pub mod net {
...
     use std::io::{self, IoSlice, IoSliceMut};
-    use std::os::fd::{AsFd, AsRawFd};
+    use std::os::fd::AsFd;
     use tokio::io::Interest;
     use tokio::net::{UdpSocket, UnixStream};
 
+    #[cfg(target_os = "linux")]
+    use nix::sys::socket;
+
+    #[cfg(target_os = "linux")]
```
Comment (cpacejo) [inline zpr-ext/src/tokio.rs]: Both `linux` and `android` here to match the use sites below.

(For context I've been specifying both of those to match what I've seen in some other libraries.  Since Android essential "is" Linux.)

File: zpr-ext/src/std.rs [hunk 3]
```diff
@@ -255,6 +255,7 @@ pub mod os {
         }
     }
 
+    #[cfg(unix)]
```
Comment (cpacejo) [inline zpr-ext/src/std.rs]: `cfg(any(doc, unix))` ensures this shows up in documentation also.  (Which I believe gets built without any "target" options set.)

File: zpr-ext/src/std.rs [hunk 4]
```diff
@@ -265,7 +266,12 @@ pub mod os {
             use std::ptr;
             use std::vec::Vec;
 
-            #[cfg(any(doc, target_os = "android", target_os = "linux"))]
+            #[cfg(any(target_os = "android", target_os = "linux"))]
```
Comment (cpacejo) [inline zpr-ext/src/std.rs]: And then also here add `doc` (but not to the "macos" variant).

File: adapter/ph/src/sys/macos/zprtun.rs [hunk 5]
```diff
@@ -1 +1,80 @@
...
+        // SAFETY: we've now initialized this much of the slice
+        unsafe {
+            buf.advance_mut(size);
+        }
+        Ok(size)
+    }
+
+    pub fn try_send(&self, _buf: &[u8]) -> std::io::Result<usize> {
+        let idev = &*(self.0);
+        idev.send(_buf)
```
Comment (cpacejo) [inline adapter/ph/src/sys/macos/zprtun.rs]: `buf` instead of `_buf` since it's used now :)

**Review by cpacejo — CHANGES_REQUESTED** [review]
Looks good!  Just a couple minor suggestions to change

File: adapter/ph/Cargo.toml (hunk 1 above)
Comment (mkolehmainen) [reply adapter/ph/Cargo.toml:44]: Noted.  Altered comment, but keeping at 0.7.1 since that one works for us for now.

File: adapter/ph/src/sys/macos/zprtun.rs (hunk 5 above)
Comment (mkolehmainen) [reply adapter/ph/src/sys/macos/zprtun.rs]: oh good catch!

File: adapter/ph/Cargo.toml (hunk 1 above)
Comment (cpacejo) [reply adapter/ph/Cargo.toml:44]: I'm OK to check this in with 0.7.1 for just macOS but I do want to unify across OSes and we can't do that with 0.7.1.  0.6.1 is the last version that supports multiqueue and vectored I/O.

**Review by cpacejo — APPROVED** [review]
Looks good thanks!
